### ISSUE TITLE
Recent change makes Equatable protocol no longer work

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#5753]
**Describe the bug**
It seems that some recent version of Pyright implicitly adds `__hash__ = None` onto classes that implement `__eq__` but not `__hash__`. This change is useful, but I think it should not affect protocols.

**Code or Screenshots**
```python
from typing import Protocol

class Equatable(Protocol):
    def __eq__(self, o: object, /) -> bool:
        ...

a: Equatable = object()  # Should work, but is reported because `__hash__` is incompatible.
_ = a

class Hashable(Equatable, Protocol):
    def __hash__(self, /) -> int:  # Should work, but is reported because of incompatible override.
        ...
```

![image](https://github.com/microsoft/pyright/assets/15389744/ce6185d9-ebaa-418b-88fa-3affa8929401)

**VS Code extension or command-line**
Pylance v2023.8.31 Pre-Release (Pyright 1.1.323)

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, the `__hash__` check should be exempt from protocol matching. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
